plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/Stablecoin.hs

```diff
@@ -0,0 +1,385 @@
+        , scStablecoinTokenName     :: TokenName -- ^ 'TokenName' of the stablecoin
+        , scReservecoinTokenName    :: TokenName -- ^ 'TokenName' of the reservecoin
+        }
+    deriving stock (Generic, Haskell.Eq, Haskell.Show)
+    deriving anyclass (ToJSON, FromJSON)
+
+{-# INLINEABLE minReserve #-}
+-- | Minimum number of base currency coins held by the bank
+minReserve :: Stablecoin -> ConversionRate -> BankState -> BC (Ratio Integer)
+minReserve Stablecoin{scMinReserveRatio} cr BankState{bsStablecoins=SC sc} =
+    let BC r = convert cr (PC $ fromInteger sc)
+    in BC $ scMinReserveRatio * r
+
+{-# INLINEABLE reservecoinNominalPrice #-}
+-- | Price of a single reservecoin in base currency
+reservecoinNominalPrice :: Stablecoin -> BankState -> ConversionRate -> BC (Ratio Integer)
+reservecoinNominalPrice Stablecoin{scReservecoinDefaultPrice} bankState@BankState{bsReservecoins=RC rc} cr =
+    if rc /= 0
+        then
+            let BC e = equity bankState cr
+            in BC (e * R.recip (fromInteger rc))
+        else fmap fromInteger scReservecoinDefaultPrice
+
+{-# INLINEABLE stablecoinNominalPrice #-}
+-- | Price of a single stablecoin in base currency. If the banks' liabilities
+--   exceed its reserves then 'stablecoinNominalPrice' is zero.
+stablecoinNominalPrice :: BankState -> ConversionRate -> BC (Ratio Integer)
+stablecoinNominalPrice bankState@BankState{bsStablecoins=SC sc} cr =
+    let BC p = convert cr (PC $ fromInteger 1)
+        BC l = liabilities bankState cr
+    in if sc == zero
+        then BC p
+        else BC $ min p l
+
+-- | Action that can be performed on the stablecoin contract.
+data SCAction
+    = MintStablecoin (SC Integer) -- ^ Create a number stablecoins, depositing the matching amount of base currency
+    | RedeemStablecoin (SC Integer) -- ^ Destroy a number of stablecoins, releasing the matching amount of base currency
+    | MintReserveCoin (RC Integer) -- ^ Create a number of reservecoins, depositing the matching amount of base currency
+    | RedeemReserveCoin (RC Integer) -- ^ Destroy a number of reservecoins, releasing the matching amount of base currency
+    deriving stock (Generic, Haskell.Eq, Haskell.Show)
+    deriving anyclass (ToJSON, FromJSON)
+
+{-# INLINEABLE calcFees #-}
+-- | Calculate transaction fees (paid in base currency to the bank) as a
+--   fraction of the transaction's volume
+calcFees :: Stablecoin -> BankState -> ConversionRate -> SCAction -> BC (Ratio Integer)
+calcFees sc@Stablecoin{scFee} bs conversionRate = \case
+    MintStablecoin (SC i) ->
+        stablecoinNominalPrice bs conversionRate * (BC scFee) * (BC $ fromInteger i)
+    RedeemStablecoin (SC i) ->
+        stablecoinNominalPrice bs conversionRate * (BC scFee) * (BC $ fromInteger i)
+    MintReserveCoin (RC i) ->
+        reservecoinNominalPrice sc bs conversionRate * (BC scFee) * (BC $ fromInteger i)
```

Comment:
Is this right? The price of one reserve coin is based on the reserves when you mint that, but then you use the same price for all `i` coins, rather than recalculating (as you should, since there are more reserve coins now, so the price should be lower). Or maybe `reservecoinNominalPrice` should compute it for a tranche of `i`.

Reply:
When you mint `i` reserve coins you also have to deposit `i * reserveCoinPrice` so the price doesn't change as a result of the minting, you just have more reserves